processor/batchprocessor/batch_processor.go

```diff
@@ -0,0 +1,223 @@
+func newBatchProcessor(params component.ProcessorCreateParams, sender consumer.TraceConsumer, cfg *Config) *batchProcessor {
+	bp := &batchProcessor{
+		name:   cfg.Name(),
+		sender: sender,
+		logger: params.Logger,
+
+		sendBatchSize: cfg.SendBatchSize,
+		tickTime:      cfg.TickTime,
+		timeout:       cfg.Timeout,
+	}
+
+	bp.batch = newBatch(bp)
+
+	// start ticker after options loaded in
+	bp.ticker = newTicker(bp)
+	go bp.ticker.start()
+
+	return bp
+}
+
+// ConsumeTrace implements batcher as a SpanProcessor and takes the provided spans and adds them to
+// batches
+func (bp *batchProcessor) ConsumeTrace(ctx context.Context, td data.TraceData) error {
+	bp.batch.add(td)
+	return nil
+}
+
+func (bp *batchProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (bp *batchProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (bp *batchProcessor) Shutdown(context.Context) error {
+	// TODO: flush accumulated data.
+	return nil
+}
+
+func (bp *batchProcessor) sendItems(
+	itemsToProcess []data.TraceData,
+	resourceSpansCount uint32,
+	measure *stats.Int64Measure,
+) {
+	td := data.NewTraceData()
+	rss := td.ResourceSpans()
+
+	// TODO: Use resourceSpansCount to set capacity of the rss slice
+	// For now there is no way to do that with internal data structure API
+
+	for _, td := range itemsToProcess {
+		td.ResourceSpans().MoveTo(rss)
+	}
+
+	statsTags := []tag.Mutator{tag.Insert(processor.TagProcessorNameKey, bp.name)}
+	_ = stats.RecordWithTags(context.Background(), statsTags, measure.M(1))
+
+	// TODO: This process should be done in an async way, perhaps with a channel + goroutine worker(s)
+	_ = bp.sender.ConsumeTrace(context.Background(), td)
+}
+
+type batch struct {
+	parent                  *batchProcessor
+	mu                      sync.RWMutex
+	items                   []data.TraceData
+	totalResourceSpansCount uint32
+	totalSpanCount          uint32
+	lastSent                int64
+}
+
+func newBatch(parent *batchProcessor) *batch {
+	return &batch{
+		parent: parent,
+		items:  make([]data.TraceData, 0, initialBatchCapacity),
+	}
+}
+
+// add updates current batch by adding new TraceData object
+func (b *batch) add(td data.TraceData) {
+	newResourceSpansCount := td.ResourceSpans().Len()
+	if newResourceSpansCount == 0 {
+		return
+	}
+
+	var itemsToProcess []data.TraceData
+	var totalResourceSpansCount uint32
+
+	b.mu.Lock()
+	b.items = append(b.items, td)
+	totalSpanCount := b.totalSpanCount + uint32(td.SpanCount())
+	b.totalSpanCount = totalSpanCount
+	b.totalResourceSpansCount = b.totalResourceSpansCount + uint32(newResourceSpansCount)
+
+	if totalSpanCount >= b.parent.sendBatchSize {
+		itemsToProcess, totalResourceSpansCount = b.getAndReset()
+	}
+	b.mu.Unlock()
```

Comment:
why not appending directly to one td? Which means 

`b.items` is `data.TraceData` not `[]data.TraceData`?

Reply:
My intention was that instead of growing the ResourceSpans slice, it might be better to use smaller slice of TraceData and at the end, before sending out, create ResourceSpans slice with already known length. I followed the same approach that was implemented before. But might be premature optimization

Reply:
Why do you think that is better? Before we had that because we needed to do batching at Node + Resource. Now we just add to the slice?

Reply:
I don't think that was done only because of Node + Resource grouping. Here https://github.com/open-telemetry/opentelemetry-collector/blob/master/processor/batchprocessor/node_batcher.go#L172 we could easily have `[]*tracepb.Span` instead of `[][]*tracepb.Span` .

I'm thinking it would be better just because we would not need to re-allocate the `ResourceSpans` slice while growing it on the fly. But the internal data interface doesn't allow to set slice capacity right now, so it doesn't really matter.

I can switch to growing ResourceSpans on the fly if I can't prove this approach with benchmarks first.

Reply:
Unless we have a proof of better performance I would recommend to choose the approach that results in simpler code.